### PR TITLE
[PY] feat: SSO - Deduplication

### PR DESCRIPTION
**Pull request by BMS-geodev**
## Linked issues

closes: #1873

## Details

This PR focuses on the sso_dialog operations. Specifically dialogs use a 2 step waterfall process to manage token acquisition and deduplication. These features generally exist in the JS version of this process, but are handled pretty differently there. 

Some tests are written to provide basic coverage of the file sso_dialog.py. 

#### Change details

Extended the file sso_dialog.py

Added test_sso_dialog.py

General fmt, lint and option config adjustments.

**code snippets**:

**screenshots**:

## Attestation Checklist

- [x] My code follows the style guidelines of this project

- I have checked for/fixed spelling, linting, and other errors
- I have commented my code for clarity
- I have made corresponding changes to the documentation (updating the doc strings in the code is sufficient)
- My changes generate no new warnings
- I have added tests that validates my changes, and provides sufficient test coverage. I have tested with:
  - Local testing
  - E2E testing in Teams
- New and existing unit tests pass locally with my changes

### Additional information

> Feel free to add other relevant information below

### CODE REVIEW

**Comment by BMS-geodev** [comment]
Need to find the equivelent of tokenExchangeOperationName from TeamsSsoBotAuthentication.ts

```
        const activity = context.activity;
        const channelId = activity.channelId;
        const conversationId = activity.conversation.id;
        if (activity.type !== ActivityTypes.Invoke || activity.name !== tokenExchangeOperationName) {
            throw new Error('TokenExchangeState can only be used with Invokes of signin/tokenExchange.');
        }
        ```

**Comment by BMS-geodev** [comment]
hmm

https://github.com/microsoft/botbuilder-python/blob/main/libraries/botbuilder-schema/botbuilder/schema/_sign_in_enums.py

```
class SignInConstants(str, Enum):
    # Name for the signin invoke to verify the 6-digit authentication code as part of sign-in.
    verify_state_operation_name = "signin/verifyState"
    # Name for signin invoke to perform a token exchange.
    token_exchange_operation_name = "signin/tokenExchange"
    # The EventActivity name when a token is sent to the bot.
    token_response_event_name = "tokens/response"
```

**Comment by BMS-geodev** [comment]
the __init__ of botbuilder.schema declares this the right way. Made more sense to see how it is materialized in the .venv

**Comment by BMS-geodev** [comment]
still need to figure out how step_two should influence the __init__ as hinted by

`# TODO - duplicate token exchange afterturn state `

**Comment by BMS-geodev** [comment]
I think I have all the concepts and TODOs covered. Gonna do a bulk fmt/lint to align.

**Comment by BMS-geodev** [comment]
trying a callback to handle:

teams/auth/sso/sso_dialog.py:47:8: E1101: Instance of 'SsoDialog' has no 'after_turn' member (no-member)

where 
```
class SsoDialog(Generic[StateT], Dialog[StateT], AuthComponent[StateT]):
    "handles dialog sso authentication"

    def __init__(self, name: str, options: SsoOptions, msal: ConfidentialClientApplication) -> None:
        super().__init__(SsoDialog.__name__)
        self.add_dialog(SsoPrompt("TeamsSsoPrompt", name, options, msal))
        self.add_dialog(
            WaterfallDialog(
                SSO_DIALOG_ID,
                [self._step_one, self._step_two],
            )
        )

        self._options = options
        self.initial_dialog_id = SSO_DIALOG_ID

        self.after_turn(self._handle_duplicate_token_exchange)

    async def _handle_duplicate_token_exchange(self, _context: TurnContext, state: StateT) -> bool:
        return not state.temp.duplicate_token_exchange
        ```
        was flagging that after_turn was a no-member.

**Comment by BMS-geodev** [comment]
on discussion with team, it seems the todo comment related to after turn was just a head's up - thought starter, not a distinct requirement. 

seen here, in TeamsSsoBotAuthentication.ts

```
export class TeamsSsoBotAuthentication<TState extends TurnState> extends BotAuthenticationBase<TState> {
    private _prompt: TeamsSsoPrompt;
    private _tokenExchangeIdRegex: RegExp;

    /**
     * Initializes a new instance of the TeamsSsoBotAuthentication class.
     * @param {Application<TState>} app - The application object.
     * @param {TeamsSsoSettings} settings - The settings for Teams SSO.
     * @param {string} settingName - The name of the setting.
     * @param {ConfidentialClientApplication} msal - The MSAL (Microsoft Authentication Library) object.
     * @param {Storage} storage - The storage object for storing state.
     */
    public constructor(
        app: Application<TState>,
        settings: TeamsSsoSettings,
        settingName: string,
        msal: ConfidentialClientApplication,
        storage?: Storage
    ) {
        super(app, settingName, storage);

        this._prompt = new TeamsSsoPrompt('TeamsSsoPrompt', settingName, settings, msal);
        this._tokenExchangeIdRegex = new RegExp(
            `[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}-${this._settingName}`
        );

        // Do not save state for duplicate token exchange events to avoid eTag conflicts
        app.turn('afterTurn', async (context, state) => {
            return state.temp.duplicateTokenExchange !== true;
        });
    }
```
the afterTurn step must ensure we done save a duplicate state. 

It is my current belief that these steps from the PR content on sso_dialog.py cover the need. specifically in _step_two and _should_dedupe.

```
    async def _step_two(self, context: WaterfallStepContext) -> DialogTurnResult:
        token_response = context.result

        if token_response and await self._should_dedup(context.context):
            context.state.temp.duplicate_token_exchange = True
            return DialogTurnResult(DialogTurnStatus.Waiting)

        context.state.temp.duplicate_token_exchange = False
        return await context.end_dialog(token_response)

    async def _should_dedup(self, context: TurnContext) -> bool:
        """
        Checks if deduplication should be performed for token exchange.
        """
        etag = context.activity.value.get("id")
        store_item = {"eTag": etag}
        key = self._get_storage_key(context)

        try:
            await self._options.storage.write({key: store_item})
        except Exception as e:
            if "eTag conflict" in str(e):
                return True
            raise e

        return False
```

File: python/packages/ai/teams/auth/sso/sso_dialog.py [hunk 1]
```diff
@@ -74,5 +76,50 @@ async def _step_one(self, context: WaterfallStepContext) -> DialogTurnResult:
     async def _step_two(self, context: WaterfallStepContext) -> DialogTurnResult:
         token_response = context.result
 
-        # TODO: Dedup token exchange responses
+        if token_response and await self._should_dedup(context.context):
+            context.state.temp.duplicate_token_exchange = True
+            return DialogTurnResult(DialogTurnStatus.Waiting)
+
+        context.state.temp.duplicate_token_exchange = False
```
Comment (singhk97) [inline python/packages/ai/teams/auth/sso/sso_dialog.py:83]: Do you know where this field is being used?

**Review by singhk97 — APPROVED** [review]

File: python/packages/ai/teams/auth/sso/sso_dialog.py [hunk 2]
```diff
@@ -74,5 +76,50 @@ async def _step_one(self, context: WaterfallStepContext) -> DialogTurnResult:
...
+        Gets the storage key for storing the token exchange state.
+        """
+        if not context or not context.activity or not context.activity.conversation:
+            raise ValueError("Invalid context, cannot get storage key!")
+
+        activity = context.activity
+        if (
+            activity.type != ActivityTypes.invoke
+            or activity.name != SignInConstants.token_exchange_operation_name
+        ):
```
[the remark this answers is not in the document]
Comment (BMS-geodev) [reply python/packages/ai/teams/auth/sso/sso_dialog.py:114]: I believe the first aligns closest to the equivalent expression in the original JS, but the latter suggestion is more pythonic. I'll run a test to ensure behavior is the same.

File: python/packages/ai/teams/auth/sso/sso_dialog.py (hunk 1 above)
Comment (lilyydu) [reply python/packages/ai/teams/auth/sso/sso_dialog.py:83]: also curious about this

File: python/packages/ai/teams/auth/sso/sso_dialog.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (BMS-geodev) [reply python/packages/ai/teams/auth/sso/sso_dialog.py]: tested and accepted the change

File: python/packages/ai/teams/auth/sso/sso_dialog.py (hunk 1 above)
Comment (BMS-geodev) [reply python/packages/ai/teams/auth/sso/sso_dialog.py:83]: ok reviewed this. 

First, its basic existance is derived from line 134 of the JS [TeamsSsoBotAuthentication.ts](https://github.com/BMS-geodev/teams-ai/blob/adc7bc5c131e8d9c211b3334717510f80aac78af/js/packages/teams-ai/src/authentication/TeamsSsoBotAuthentication.ts#L134) 

it is used in the afterturn concept that the TS explicitly discusses. As we've reviewed elsewhere, the python implementation does not express the afterturn task the same way. 

it is touched last here:

```
        app.turn('afterTurn', async (context, state) => {
            return state.temp.duplicateTokenExchange !== true;
        });
```
So, I'm not sure that we need it and will test today.

Comment (BMS-geodev) [reply python/packages/ai/teams/auth/sso/sso_dialog.py:83]: test passing when it is removed. 

As we've determined we do not need the same afterturn expression (in the python implementation), then this explicit bool flag is probably not required either. 

I will remove it.

@singhk97 @lilyydu

**Comment by BMS-geodev** [comment]
addressed all comments, test, fmt, lint working locally

**Review by lilyydu — APPROVED** [review]

**Pull request merged**